core/src/main/kotlin/com/justai/jaicf/api/routing/BotRoutingApi.kt

```diff
@@ -0,0 +1,164 @@
+    }
+
+    /**
+     * Route current bot request back to previous bot engine
+     *
+     * @throws BotRequestRerouteException to reroute request using [BotRoutingEngine]
+     * @return [Nothing] as no request execution possible after invoking this method
+     *
+     * @see BotRoutingEngine
+     * */
+    fun routeBack(): Nothing {
+        try {
+            val routingContext = botContext.routingContext
+            val curr = routingContext.routingEngineStack.pop()
+            val target = routingContext.routingEngineStack.pop()
+            logger.info("Routing request back from engine: $curr to engine: $target")
+            throw BotRequestRerouteException(target)
+        } catch (e: EmptyStackException) {
+            logger.warn("Failed to change route back as there is no engines left in stack")
+            throw NoRouteBackException
+        }
+    }
+
+    /**
+     * Route client all next requests to specified [engineName].
+     *
+     * @param engineName target engine name specified in routables map in [BotRoutingEngine]
+     * @param targetState target state for scenario in [engineName].
+     *
+     * @see BotRoutingEngine
+     * */
+    fun changeEngine(engineName: String, targetState: String? = null) {
+        targetState?.let {
+            val dialogContext = botContext.routingContext.dialogContextMap.getOrDefault(engineName, DialogContext())
+            val currentState = StatePath.parse(dialogContext.currentState)
+            val resolved = currentState.resolve(targetState).toString()
+            botContext.routingContext.targetState = resolved
+        }
+
+        botContext.routingContext.routingEngineStack.push(engineName)
+    }
+
+    /**
+     * Route client next requests back to previous bot engine
+     *
+     * @see BotRoutingEngine
+     * */
+    fun changeEngineBack() {
+        try {
+            val routingContext = botContext.routingContext
+            val curr = routingContext.routingEngineStack.pop()
+            val target = routingContext.routingEngineStack.peek()
+            logger.info("Changing execution back from engine: $curr to engine: $target")
+        } catch (e: EmptyStackException) {
+            logger.warn("Failed to change bot engine back as there is no engines left in stack")
+        }
+    }
+}
+
+/**
+ * A context used in [BotRoutingEngine] and [BotRoutingApi] to store routing state.
+ *
+ * @see BotRoutingApi
+ * @see BotRoutingEngine
+ * */
+internal data class BotRoutingContext(
+    val dialogContextMap: MutableMap<String, DialogContext> = mutableMapOf(),
+    val routingEngineStack: Stack<String> = Stack(),
+    var targetState: String? = null,
+    var currentEngine: String = BotRoutingEngine.DEFAULT_ROUTE_NAME,
+)
+
+/**
+ * A helpful extension to get [BotRoutingContext] from BotEngine.
+ *
+ * @see BotRoutingContext
+ * @see BotRoutingApi
+ * @see BotRoutingEngine
+ * */
+internal val BotContext.routingContext: BotRoutingContext
```

Comment:
Why internal?

Reply:
General user should not see it. Anyone who wants to modify this context can copy-paste deserialization logic form sources.

Reply:
A user may want to read it to build its own logic. You can make public getters and internal or private mutators